db/commitlog/commitlog.cc

```diff
@@ -188,14 +189,54 @@ db::commitlog::descriptor::operator db::replay_position() const {
     return replay_position(id);
 }
 
+/**
+ * virtual dispatch for actually inputting data.
+ * purposely de/un-templated
+ *
+ * Writes N entries to a single segment,
+ * where each entry has its own header+crc,
+ * i.e. will be deserialized separately.
+ */
 struct db::commitlog::entry_writer {
     force_sync sync;
-    explicit entry_writer(force_sync sync_) : sync(sync_) {}
+    size_t num_entries;
+
+    explicit entry_writer(force_sync fs, size_t ne = 1)
+        : sync(fs)
+        , num_entries(ne)
+    {}
+    virtual ~entry_writer() = default;
+
+    /** return the CF id for n:th entry */
+    virtual const cf_id_type& id(size_t) const = 0;
+
+    /**
+     * Returns segment-independent size of all entries combined. Must be >= than segment-dependant total size.
+     * This is always called first, and should return "worst-case"
+     * for the complete set of entries
+     */
+    virtual size_t size() const = 0;
+    /**
+     * Return the total size of all entries in this given segment
+     * Called after size(void), once a segment has been chosen.
+     * Should return the total, exact, size for all entries + overhead (i.e. schema)
+     * for this segment.
+     *
+     * Can be called more than once, if segment switch is neccesary (because race)
+     */
     virtual size_t size(segment&) = 0;
-    // Returns segment-independent size of the entry. Must be <= than segment-dependant size.
-    virtual size_t size() = 0;
-    virtual void write(segment&, output&) = 0;
-    virtual ~entry_writer() {};
+    /**
+     * return the size of the n:th entry in this given segment
+     * Only called IFF num_entries > 1, and if so, after size(size_t)
```

Comment:
What is `size(size_t)`? Did you mean `size()`?

Reply:
I did.

Reply:
fixed.